### DISCUSSION TITLE
"Document Type" field on .docx file isn't being saved to SharePoint

### DISCUSSION THREAD

**Discussion by aaronmorgan** [pnp/pnpcore#1060] [Q&A]
PnP.Core & PnP.Core.Auth, version: 1.8.38-nightly

I've got the most odd problem; a .docx file I created for testing (contains only some filler text) will not have one of it's fields set when a file in SharePoint. 

In fact there's weird behaviour for Choices and Managed Metadata columns. 

I upload a file like this:

```
IFile file = await documentsFolder.Files.AddAsync(
  name: parameter.Filename,
  content: fileStream,
  overwrite: true);
```

...and then set properties from a dictionary like this:

```
await file.LoadAsync(p => p.ListItemAllFields);

foreach (var property in parameter.Properties)
{
  file.ListItemAllFields[property.Key] = property.Value;
}

file.ListItemAllFields[SharePointFields.CONTENT_TYPEID] = parameter.ContentTypeId;
await file.ListItemAllFields.UpdateOverwriteVersionAsync();
```

If the file is a .docx the `Document_x0020_Type` column is not set, it's always null. Despite if I debug the code, the File's ListItemAllFields having the property set before the update is called. 
Other columns and Choice fields are set correctly.

![image](https://user-images.githubusercontent.com/4121193/207573444-70c42f6d-e190-4aed-b303-935f9a49d857.png)

If the file is a PDF it works just fine.

I tried using some of the other update methods, e.g. `UpdateAsync`, and that made the problem worse because then none of the Choice fields were set either.

I've gone round and round on this and while other columns are working fine there's some weird interplay between the .docx file and Choice and Managed Metadata columns.

Have you seen or heard of anything like this? 
What other information could I provide to help with understanding the cause?

[offer_letter.docx](https://github.com/pnp/pnpcore/files/10227031/offer_letter.docx)

Thanks,
Aaron.

**Comment by jansenbe** [comment]
@aaronmorgan : can you try to load the relevant list information when you load `ListItemAllFields` (see https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#updatingreading-file-metadata for details). Internally we need to have the the list fields loaded to correctly identify the type of field. When possible this data is loaded automatically, but that's not always possible.

**Reply by aaronmorgan** [reply]
Thank you. I've changed the code to the following but the problem exists. It does certainly gel with the comment in the docs, not loading the parent list correctly can result in metadata updates not working. 

I modfied the example to load `All` properties for the parent folder too.

```
IFile file = await documentsFolder.Files.AddAsync(
  name: parameter.Filename,
  content: fileStream,
  overwrite: true);

var documentUrl = $"{documentsFolder.ServerRelativeUrl}/{parameter.Filename}";

file = await pnpContext.Web.GetFileByServerRelativeUrlAsync(documentUrl,
  f => f.ListItemAllFields.QueryProperties(li => li.All,
    li => li.ParentList.QueryProperties(p => p.All,
      p => p.Fields.QueryProperties(p => p.All))));
```

I inspected the properties of the IFile after getting it and the DocumentType is there, so it's read it ok it seems.

![image](https://user-images.githubusercontent.com/4121193/207717286-777387ae-1fbf-43ff-822d-7f51cd365bb9.png)

Am I fetching the file correctly?

Thanks,
Aaron.

**Reply by jansenbe** [reply]
@aaronmorgan : In my testing I don't see the "Document_x0020_Type" column on document libraries, I assume it's a custom column on your site. Can you share more details on that column? So the issue is that you can't update that column's metadata whenever the uploaded document is a word doc, but it works for PDF?

**Reply by aaronmorgan** [reply]
Hi Bert, that's correct .docx Word files are not updating, PDF files do.

It is a custom column, and the type is 'Managed Metadata'.

The Term Set Settings field is configured to point to a term set, e.g.

![image](https://user-images.githubusercontent.com/4121193/207831758-3ce69fd6-c41f-4b7a-aea6-557f19b74840.png)

Allow 'Fill-in' choices is true and there's no Default Value configured.

Aaron.

**Reply by jansenbe** [reply]
And the term you want to set for the word document does exist in the termset? The APIs will not automatically add terms when they're not present (at least that's my thinking). UI can do that. Can you  verify if the document type you need for Word is in the term set and then test again. Also updating managed metadata fields in PnP Core is different from setting text fields, see https://pnp.github.io/pnpcore/using-the-sdk/listitems-fields.html#taxonomy-fields

**Reply by aaronmorgan** [reply]
Hi, yes good question. Yes the term exists, I've double checked the Guid id against the SharePoint UI and it's the same term used on the .pdf files which does work.

Aaron.

**Reply by jansenbe** [reply]
Can you trim the meta data update to just this field to rule out the impact of any other fields?

**Reply by aaronmorgan** [reply]
Hi, sorry following up on this I have some new information. 

In my testing I'm using the same binary file, an actual .docx document with minimal body text. If I rename it to .doc or .pdf the Document Type is set fine. 

Document Type is a Term loaded from a global term store and it and other terms in the same Term Group work fine on any other file type I've tested.

When the problem manifests on a .docx file the Content Type is also incorrect, it's Document when it should be a custom type derived from MasterDocument

Here's the code in it's entirety, it's pretty much what's in the documentation examples for "Working with Files | Updating the file Author, Editor, Created or Modified properties".

Setup:
```
IFile file = await documentsFolder.Files.AddAsync(
    name: parameter.Filename,
    content: fileStream,
    overwrite: true);

await SetFileMetadata(parameter, file);
```

Execution:
```
await file.LoadAsync(p => p.ListItemAllFields);

foreach (var property in parameter.Properties)
{
    file.ListItemAllFields[property.Key] = property.Value;
}

file.ListItemAllFields[SharePointFields.CONTENT_TYPEID] = parameter.ContentTypeId;

await file.ListItemAllFields.UpdateOverwriteVersionAsync();
```

I tried as you suggested and set only the affected field, same problem.

I did notice once when trying to set the field manually through the Item properties 'Edit All' pane that on saving the new tag value this error occurred; "The file is locked for shared use by..." making me wonder if there's a race condition on these file types. 

![image](https://user-images.githubusercontent.com/4121193/224677645-37cf8d71-7485-4ef8-884b-76b2dc6e7f1b.png)

Repeated attempts to reproduce this have failed and the property can be set manually using the 'Edit All' option. Though this won't work for my purpose as we're bulk migration millions of documents that need this field set :)

Searching Google it does seem newer Word documents can act oddly in SharePoint, particularly when it comes to setting custom properties. It's like SharePoint assumes something about it's metadata that's in conflict with the setting of custom properties in PnP Core.

**Reply by jansenbe** [reply]
@aaronmorgan : did you experiment with adding a wait time after the file add and setting it's properties?

**Reply by aaronmorgan** [reply]
Ok, good idea; just now tried `Thread.Sleep(10000)` and same behaviour. Also using `overwrite: false`. The custom property still isn't set.

**Reply by aaronmorgan** [reply]
I've also upgraded to the latest nightly incase that helped.

**Reply by jansenbe** [reply]
@aaronmorgan : and if you leave out the content type update or first commit the content type update and then handle the fields coming with the content 
type

**Reply by aaronmorgan** [reply]
Thanks Bert, that didn't affect the behaviour either. 

I heard from one of our site testers that .xlsm files exhibit the same problem. This gels with the issue somehow being centered around newer Office documents.

We need to get to prod with this so I've worked around the problem by uploading the file with a `.tmp` extension to the filename, and simply remove it after setting the meta data and everything is fine.

Thanks for your help in trying to understand and resolve this for me. If I learn more about it I'll post back here.
